### PR TITLE
BUPH-143 | fabrication file writer

### PR DESCRIPTION
**Pull request by jpmarcotte**
Option 1 for OO Buphalo file generation. See the tests for usage.

PROs:
- Uses existing Buphalo objects
- Is automatically included in every version of Buphalo

CONs:
- Because it uses existing Buphalo objects, there's a lot of unnecessary fields that don't need to be set for this use-case and will be ignored
- A separate Writer is necessary because dumping the objects would just leave too much confusing extra stuff

### CODE REVIEW

**Review by mucha55 — COMMENTED** [review]
Definitely prefer the dedicated models in #75 over this, but this approach has some nice things that would be difficult to work into the other options. What jumped out at me was the `\` prefix for FQCNs; with option 3 userspace would need to a) know to do that, and b) do it, and I'm sure there will be other little things like that in the future

**Comment by jpmarcotte** [comment]
> What jumped out at me was the \ prefix for FQCNs;

Yeah, I'm still not entirely sure of the necessity of it. My understanding is that it is a micro-optimization in terms of name-resolution (I know that's the case with many built-in php functions), but I'm not sure if that's the case with class names or not. If anything, that's probably something we could just have Buphalo do when creating an AP instead of relying on clients to do it.

i.e.
```php
$annotationProcessorFQCN =
    ($annotationProcessorFQCN[0] === '\')
        ? $annotationProcessorFQCN
        : '\' . $annotationProcessorFQCN;
return new $annotationProcessorFQCN();
```

**Comment by jpmarcotte** [comment]
Some quick tests (3 iterations of 100 runs of that test, after 1 burning run to potentially prime opcache) `time for i in {1..100}; do ./tests/v1/ApiDumper/run_test; done` ordered in increasing time. 
Without slashes: 17.620s, 17.751s, 18.390s
With slashes:  17.338s, 17.550s, 17.929s

Difference is slightly lower (~2%), but the ranges overlap, so... _shrug_

**Comment by jpmarcotte** [comment]
Also just ran this test:
```
$ cat ./test-with-slashes.php 
<?php
require_once('./vendor/autoload.php');
$ap = new \Neighborhoods\Buphalo\V1\AnnotationProcessors\EmptyString();
exit();
$ time for i in {1..1000}; do php ./test-with-slashes.php; done;

real    0m46.736s
user    0m28.679s
sys     0m12.938s

$ cat ./test-without-slashes.php 
<?php
require_once('./vendor/autoload.php');
$ap = new Neighborhoods\Buphalo\V1\AnnotationProcessors\EmptyString();
exit();

$ time for i in {1..1000}; do php ./test-without-slashes.php; done;

real    0m45.916s
user    0m28.515s
sys     0m12.637s
```

So yeah, I'm thinking it doesn't make enough of a difference to care about, if there is any at all.

**Comment by jpmarcotte** [comment]
> I'm sure there will be other little things like that in the future

This could also still be handled by the dedicated Dumper in #75.

File: tests/v1/FabricationFileWriter/run_test [hunk 1]
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env php
+<?php
+declare(strict_types=1);
+error_reporting(E_ALL);
+
+use Neighborhoods\Buphalo\V1\FabricationFile;
+use Neighborhoods\Buphalo\V1\AnnotationProcessor;
+use Neighborhoods\Buphalo\V1\AnnotationProcessors;
+
+$PWD = __DIR__;
```
Comment (arielallon) [inline tests/v1/FabricationFileWriter/run_test:10]: same nittiest nit as in #75

File: src/V1/FabricationFile/WriterInterface.php [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neighborhoods\Buphalo\V1\FabricationFile;
+
+use Neighborhoods\Buphalo\V1\FabricationFileInterface;
+
+interface WriterInterface
```
Comment (arielallon) [inline src/V1/FabricationFile/WriterInterface.php:9]: fwiw, i like `Writer` better than `Dumper`, but it's totally personal taste

**Review by arielallon — APPROVED** [review]
also good stuff

i don't think i have a strong opinion between this and option 3, though i agree a leaning towards option 3 as more pit of success-y

**Comment by jpmarcotte** [comment]
Opting for #75 instead.

**Pull request closed without merging**